core/src/retransmit_stage.rs

```diff
@@ -361,74 +297,56 @@ pub fn retransmitter(
         CLUSTER_NODES_CACHE_NUM_EPOCH_CAP,
         CLUSTER_NODES_CACHE_TTL,
-    ));
-    let hasher_reset_ts = Arc::default();
-    let stats = Arc::new(RetransmitStats::default());
-    let shreds_received = Arc::new(Mutex::new((
-        LruCache::new(DEFAULT_LRU_SIZE),
-        PacketHasher::default(),
-    )));
-    let first_shreds_received = Arc::new(Mutex::new(BTreeSet::new()));
-    (0..sockets.len())
-        .map(|s| {
-            let sockets = sockets.clone();
-            let bank_forks = bank_forks.clone();
-            let leader_schedule_cache = leader_schedule_cache.clone();
-            let shreds_receiver = shreds_receiver.clone();
-            let cluster_info = cluster_info.clone();
-            let stats = stats.clone();
-            let cluster_nodes_cache = Arc::clone(&cluster_nodes_cache);
-            let hasher_reset_ts = Arc::clone(&hasher_reset_ts);
-            let shreds_received = shreds_received.clone();
-            let max_slots = max_slots.clone();
-            let first_shreds_received = first_shreds_received.clone();
-            let rpc_subscriptions = rpc_subscriptions.clone();
-
-            Builder::new()
-                .name("solana-retransmitter".to_string())
-                .spawn(move || {
-                    trace!("retransmitter started");
-                    loop {
-                        if let Err(e) = retransmit(
-                            &bank_forks,
-                            &leader_schedule_cache,
-                            &cluster_info,
-                            &shreds_receiver,
-                            &sockets[s],
-                            s as u32,
-                            &stats,
-                            &cluster_nodes_cache,
-                            &hasher_reset_ts,
-                            &shreds_received,
-                            &max_slots,
-                            &first_shreds_received,
-                            rpc_subscriptions.as_deref(),
-                        ) {
-                            match e {
-                                Error::RecvTimeout(RecvTimeoutError::Disconnected) => break,
-                                Error::RecvTimeout(RecvTimeoutError::Timeout) => (),
-                                _ => {
-                                    inc_new_counter_error!("streamer-retransmit-error", 1, 1);
-                                }
-                            }
-                        }
-                    }
-                    trace!("exiting retransmitter");
-                })
-                .unwrap()
+    );
+    let mut hasher_reset_ts = Instant::now();
+    let mut stats = RetransmitStats::default();
+    let shreds_received = Mutex::new((LruCache::new(DEFAULT_LRU_SIZE), PacketHasher::default()));
+    let first_shreds_received = Mutex::<BTreeSet<Slot>>::default();
+    let num_threads = get_thread_count().min(8).max(sockets.len());
```

Comment:
Was `8` chosen to match the value in `new_with_external_ip`? Could this be a shared constant?

Reply:
I was borrowing that `.min(8)` from gossip:
https://github.com/solana-labs/solana/blob/d7051b0d2/gossip/src/cluster_info.rs#L1643-L1647

in order not to take too much cpu for one process. It does not actually need to match `sockets.len()`, and we can have more  threads than sockets.